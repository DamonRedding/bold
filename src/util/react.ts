import { MutableRefObject, Ref } from 'react'

export function composeRefs<T>(...refs: Array<Ref<T>>) {
    return (value: T) => refs.forEach(ref => setRef(ref, value))
}

export function setRef<T>(ref: Ref<T>, value: any) {
    if (typeof ref === 'function') {
        ref(value)
    } else if (ref) {
        (ref as MutableRefObject<T>).current = value
    }
}

/**
 * Returns the next sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getNextSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.nextElementSibling) {
        current = current.nextElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}

/**
 * Returns the previous sibling of `element` that matches the `predicate`.
 * Returns `null` if not found.
 */
export function getPreviousSibling(element: Element, predicate: (sibling: Element) => boolean) {
    let current: Element = element

    while (current.previousElementSibling) {
        current = current.previousElementSibling

        if (predicate(current)) {
            return current
        }
    }

    return null
}
