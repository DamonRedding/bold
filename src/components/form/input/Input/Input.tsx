import * as React from 'react'

export interface InputController {
  getInput(): HTMLInputElement
  focus(): any
  blur(): any
}

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: <T>(event?: React.FocusEvent<T>) => void
  onChange?: <T>(event: React.ChangeEvent<T> | any) => void
  onFocus?: <T>(event?: React.FocusEvent<T>) => void
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
  provideController?: (controller: InputController) => void
  autoFocus?: boolean
}

export interface InputProps extends PublicInputProps {
  type?: string
}

export class Input extends React.Component<InputProps, any> {
  private input: HTMLInputElement

  componentDidMount() {
    this.props.provideController && this.props.provideController({
      getInput: () => this.input,
      focus: () => this.input.focus(),
      blur: () => this.input.blur(),
    })
  }

  componentWillUnmount() {
    this.props.provideController && this.props.provideController(null)
  }

  render() {
    const { provideController, ...rest } = this.props

    return (
      <input ref={this.ref} {...rest} />
    )
  }

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

}
