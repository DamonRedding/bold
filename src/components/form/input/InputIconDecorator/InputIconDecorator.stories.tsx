import { action } from '@storybook/addon-actions'
import { select, withKnobs } from '@storybook/addon-knobs'
import { storiesOf } from '@storybook/react'
import * as React from 'react'

import { withPropTypes, withTheme } from '../../../../stories-addons'
import { TextInput } from '../TextInput/TextInput'

import { InputIconDecorator } from './InputIconDecorator'

const positionOptions = {
    'left': 'left',
    'right': 'right',
}

storiesOf('Form', module)
    .addDecorator(withPropTypes())
    .addDecorator(withKnobs)
    .addDecorator(withTheme())
    .add('InputIconDecorator', () => (
        <InputIconDecorator
            icon='lupa'
            position={select('position', positionOptions, 'right')}
            onClick={action('icon-clicked')}
        >
            <TextInput />
        </InputIconDecorator>
    ))
