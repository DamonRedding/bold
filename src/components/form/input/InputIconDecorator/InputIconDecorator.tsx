import * as React from 'react'

import { Styles, withStyles, WithStylesProps } from '../../../../styles'
import { Button, ButtonProps } from '../../../elements/Button'
import { Icons } from '../../../elements/Icon/generated/Icons'
import { Icon } from '../../../elements/Icon/Icon'

export interface InputIconDecoratorProps extends WithStylesProps, Pick<ButtonProps, 'onClick'> {
    icon: Icons
    position?: 'left' | 'right'
}

@withStyles
export class InputIconDecorator extends React.PureComponent<InputIconDecoratorProps> {

    static defaultProps: Partial<InputIconDecoratorProps> = {
        position: 'right',
    }

    render() {
        const { css, theme, children, icon, position, onClick } = this.props
        const styles: Styles = {
            wrapper: {
                position: 'relative',
            },
            left: {
                'input': {
                    paddingLeft: '3rem',
                },
            },
            right: {
                'input': {
                    paddingRight: '3rem',
                },
            },
            iconWrapper: {
                position: 'absolute',
                backgroundColor: theme.pallete.surface.background,
                display: 'flex',
                alignItems: 'center',
                justifyContent: 'center',
                padding: '0.25rem',
                width: '2.5rem',
                top: 1,
                bottom: 1,
            },
            iconLeft: {
                left: 1,
            },
            iconRight: {
                right: 1,
            },
            icon: {
                borderRadius: 'inherit',
                '&:focus': {
                    boxShadow: 'none',
                },
            },
        }

        const wrapperClasses = [
            styles.wrapper,
            position === 'left' && styles.left,
            position === 'right' && styles.right,
        ]
        const iconBoxClasses = [
            styles.iconWrapper,
            position === 'left' && styles.iconLeft,
            position === 'right' && styles.iconRight,
        ]

        return (
            <div className={css(wrapperClasses)}>
                {children}

                <span className={css(iconBoxClasses)}>
                    {onClick ?
                        <Button
                            size='small'
                            skin='ghost'
                            icon={icon}
                            tabIndex={-1}
                            onClick={this.props.onClick}
                            style={styles.icon}
                        /> :
                        <Icon
                            icon={icon}
                            style={styles.icon}
                        />
                    }
                </span>
            </div>
        )
    }
}
