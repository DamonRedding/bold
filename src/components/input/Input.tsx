import * as React from 'react'
import { EventOrValueHandler } from 'redux-form'
import * as classnames from 'classnames'
import { helpersClassnames, extractProps } from '../../util/Util'
import { ChangeEvent, FocusEvent } from 'react'

export interface PublicInputProps {
  className?: string
  checked?: boolean
  disabled?: boolean
  id?: string
  maxLength?: number
  name?: string
  onBlur?: EventOrValueHandler<FocusEvent<any>>
  onChange?: EventOrValueHandler<ChangeEvent<any>>
  onKeyPress?: (event: React.KeyboardEvent<HTMLInputElement>) => void
  placeholder?: string
  value?: any
}

export interface InputProps extends PublicInputProps {
  type: string
}

const propsToKeep = [
  'checked',
  'disabled',
  'id',
  'maxLength',
  'name',
  'onBlur',
  'onChange',
  'onKeyPress',
  'placeholder',
  'type',
  'value'
]

export class Input extends React.Component<InputProps, any> {
  public input: HTMLInputElement

  private ref = (elem: HTMLInputElement) => {
    this.input = elem
  }

  focus() {
    this.input.focus()
  }

  blur() {
    this.input.blur()
  }

  render() {
    const classes = classnames('input', this.props.className, helpersClassnames(this.props))

    return (
      <input ref={this.ref} {...extractProps(this.props, ...propsToKeep) } className={classes} />
    )
  }

}
