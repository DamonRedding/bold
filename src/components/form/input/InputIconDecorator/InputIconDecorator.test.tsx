import { render } from 'enzyme'
import * as moment from 'moment'
import * as React from 'react'

import { withTheme } from '../../../../test'
import { Input } from '../Input/Input'

import { InputIconDecorator } from './InputIconDecorator'

it('should render correctly', () => {
    const wrapper = render(withTheme(
        <InputIconDecorator icon='search'>
            <Input />
        </InputIconDecorator>
    ))
    expect(wrapper).toMatchSnapshot()
})

it('should render correctly with button', () => {
    const wrapper = render(withTheme(
        <InputIconDecorator icon='search' onClick={jest.fn()}>
            <Input />
        </InputIconDecorator>
    ))
    expect(wrapper).toMatchSnapshot()
})

it('should render correctly with left position', () => {
    const wrapper = render(withTheme(
        <InputIconDecorator icon='search' position='left'>
            <Input />
        </InputIconDecorator>
    ))
    expect(wrapper).toMatchSnapshot()
})
