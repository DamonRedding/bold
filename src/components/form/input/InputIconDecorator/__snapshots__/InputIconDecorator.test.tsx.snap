// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
.emotion-3 {
  position: relative;
}

.emotion-3 input {
  padding-right: 2.5rem;
}

.emotion-2 {
  position: absolute;
  background-color: #f2f2f7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem;
  top: 1px;
  bottom: 1px;
  right: 1px;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-1:focus {
  box-shadow: none;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

.emotion-0:focus {
  box-shadow: none;
}

<p
  class="emotion-3"
>
  <input />
  <span
    class="emotion-2"
  >
    <span
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        class="emotion-0"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M20.21 18.79l-5.32-5.32A5.93 5.93 0 0 0 16 10a6 6 0 1 0-6 6 5.93 5.93 0 0 0 3.47-1.11l5.32 5.32zM10 14a4 4 0 1 1 4-4 4 4 0 0 1-4 4z"
          id="Pesquisar"
        />
      </svg>
    </span>
  </span>
</p>
`;

exports[`should render correctly with button 1`] = `
.emotion-4 {
  position: relative;
}

.emotion-4 input {
  padding-right: 2.5rem;
}

.emotion-3 {
  position: absolute;
  background-color: #f2f2f7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem;
  top: 1px;
  bottom: 1px;
  right: 1px;
}

.emotion-2 {
  background-color: transparent;
  border: none;
  border-radius: 4px;
  color: #666666;
  cursor: pointer;
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  padding: 0;
  position: relative;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: all .2s;
  transition: all .2s;
  -webkit-transform: translate3d(0,0,0);
  -ms-transform: translate3d(0,0,0);
  transform: translate3d(0,0,0);
}

.emotion-2 span {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
  -webkit-transition: color .2s;
  transition: color .2s;
}

.emotion-2:disabled {
  opacity: 0.5;
}

.emotion-2:focus {
  outline: none;
  box-shadow: 0 0 0 2px #056DFF;
}

.emotion-2:focus {
  box-shadow: none;
}

.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

<p
  class="emotion-4"
>
  <input />
  <span
    class="emotion-3"
  >
    <button
      class="emotion-2"
      tabindex="-1"
      type="button"
    >
      <span>
        <span
          class="emotion-1"
        >
          <svg
            aria-hidden="true"
            class="emotion-0"
            height="1em"
            viewBox="0 0 24 24"
            width="1em"
          >
            <path
              d="M20.21 18.79l-5.32-5.32A5.93 5.93 0 0 0 16 10a6 6 0 1 0-6 6 5.93 5.93 0 0 0 3.47-1.11l5.32 5.32zM10 14a4 4 0 1 1 4-4 4 4 0 0 1-4 4z"
              id="Pesquisar"
            />
          </svg>
        </span>
      </span>
    </button>
  </span>
</p>
`;

exports[`should render correctly with left position 1`] = `
.emotion-1 {
  display: -webkit-inline-box;
  display: -webkit-inline-flex;
  display: -ms-inline-flexbox;
  display: inline-flex;
}

.emotion-1:focus {
  box-shadow: none;
}

.emotion-0 {
  fill: currentColor;
  font-size: 1.5rem;
}

.emotion-0:focus {
  box-shadow: none;
}

.emotion-3 {
  position: relative;
}

.emotion-3 input {
  padding-left: 2.5rem;
}

.emotion-2 {
  position: absolute;
  background-color: #f2f2f7;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 0.25rem;
  top: 1px;
  bottom: 1px;
  left: 1px;
}

<p
  class="emotion-3"
>
  <input />
  <span
    class="emotion-2"
  >
    <span
      class="emotion-1"
    >
      <svg
        aria-hidden="true"
        class="emotion-0"
        height="1em"
        viewBox="0 0 24 24"
        width="1em"
      >
        <path
          d="M20.21 18.79l-5.32-5.32A5.93 5.93 0 0 0 16 10a6 6 0 1 0-6 6 5.93 5.93 0 0 0 3.47-1.11l5.32 5.32zM10 14a4 4 0 1 1 4-4 4 4 0 0 1-4 4z"
          id="Pesquisar"
        />
      </svg>
    </span>
  </span>
</p>
`;
